File: ios/App/Sources/Core/Domain/NavCaddy/Navigation/NavigationExecutor.swift
import Foundation

/// Processes RoutingResult and executes navigation actions.
///
/// Spec R3: Executes navigation as deep link with <100ms latency budget.
/// Converts routing decisions into navigation coordinator actions.
@MainActor
struct NavigationExecutor {
    private let coordinator: NavigationCoordinator
    private let deepLinkBuilder: DeepLinkBuilder

    init(coordinator: NavigationCoordinator, deepLinkBuilder: DeepLinkBuilder = DeepLinkBuilder()) {
        self.coordinator = coordinator
        self.deepLinkBuilder = deepLinkBuilder
    }

    /// Executes a routing result
    ///
    /// - Parameter result: The routing result from RoutingOrchestrator
    /// - Returns: NavigationAction describing what UI action should be taken
    func execute(_ result: RoutingResult) -> NavigationAction {
        switch result {
        case .navigate(let target, let intent):
            return executeNavigation(target: target, intent: intent)

        case .noNavigation(let intent, let response):
            // No navigation needed, just show response
            return .showResponse(response: response, intent: intent)

        case .prerequisiteMissing(let intent, let missing, let message):
            // Prerequisites are missing, show message and optionally navigate to setup
            return .showPrerequisitePrompt(
                message: message,
                prerequisites: missing,
                intent: intent
            )

        case .confirmationRequired(let intent, let message):
            // Need user confirmation before proceeding
            return .requestConfirmation(message: message, intent: intent)
        }
    }

    /// Executes navigation for a valid routing target
    private func executeNavigation(target: RoutingTarget, intent: ParsedIntent) -> NavigationAction {
        // Validate target first
        guard deepLinkBuilder.validate(target) else {
            return .showError(
                message: "Invalid navigation target for \(intent.intentType.rawValue)",
                intent: intent
            )
        }

        // Build destination from target
        guard let destination = deepLinkBuilder.buildDestination(from: target) else {
            return .showError(
                message: "Could not build destination for \(target.screen)",
                intent: intent
            )
        }

        // Execute navigation
        coordinator.navigate(to: destination)

        // Return success action
        return .navigated(destination: destination, intent: intent)
    }

    /// Navigates back
    func navigateBack() {
        coordinator.navigateBack()
    }

    /// Pops to root
    func popToRoot() {
        coordinator.popToRoot()
    }

    /// Replaces current destination
    func replace(with destination: NavCaddyDestination) {
        coordinator.replace(with: destination)
    }
}

/// Describes the action that should be taken by the UI layer after processing a routing result
enum NavigationAction {
    /// Navigation was executed successfully
    case navigated(destination: NavCaddyDestination, intent: ParsedIntent)

    /// Show a response without navigation
    case showResponse(response: String, intent: ParsedIntent)

    /// Show a prerequisite prompt with missing items
    case showPrerequisitePrompt(message: String, prerequisites: [Prerequisite], intent: ParsedIntent)

    /// Request user confirmation before proceeding
    case requestConfirmation(message: String, intent: ParsedIntent)

    /// Show an error message
    case showError(message: String, intent: ParsedIntent)
}

extension NavigationAction {
    /// The intent associated with this action
    var intent: ParsedIntent {
        switch self {
        case .navigated(_, let intent),
             .showResponse(_, let intent),
             .showPrerequisitePrompt(_, _, let intent),
             .requestConfirmation(_, let intent),
             .showError(_, let intent):
            return intent
        }
    }

    /// User-facing message, if applicable
    var message: String? {
        switch self {
        case .navigated:
            return nil
        case .showResponse(let response, _):
            return response
        case .showPrerequisitePrompt(let message, _, _):
            return message
        case .requestConfirmation(let message, _):
            return message
        case .showError(let message, _):
            return message
        }
    }

    /// Whether this action requires showing UI feedback
    var requiresUIFeedback: Bool {
        switch self {
        case .navigated:
            return false // Navigation itself is visual feedback
        case .showResponse, .showPrerequisitePrompt, .requestConfirmation, .showError:
            return true
        }
    }
}

File: android/app/src/main/java/com/example/app/domain/navcaddy/navigation/NavigationExecutor.kt
package caddypro.domain.navcaddy.navigation

import caddypro.domain.navcaddy.routing.RoutingResult
import javax.inject.Inject

/**
 * Executes navigation actions based on RoutingResult from the orchestrator.
 *
 * Processes the routing decision and:
 * - For Navigate: builds the route and executes navigation
 * - For NoNavigation: returns the response for inline display
 * - For PrerequisiteMissing: returns UI action to prompt user
 * - For ConfirmationRequired: returns UI action to show confirmation
 *
 * Respects the latency budget of <100ms for navigation execution (R3, C1).
 *
 * Spec reference: navcaddy-engine.md R3, navcaddy-engine-plan.md Task 11
 */
class NavigationExecutor @Inject constructor(
    private val deepLinkBuilder: DeepLinkBuilder,
    private val navigator: NavCaddyNavigator
) {

    /**
     * Execute the appropriate action based on routing result.
     *
     * @param result Routing decision from orchestrator
     * @return NavigationAction describing what happened (for UI feedback)
     */
    fun execute(result: RoutingResult): NavigationAction {
        return when (result) {
            is RoutingResult.Navigate -> {
                executeNavigation(result)
            }

            is RoutingResult.NoNavigation -> {
                NavigationAction.ShowInlineResponse(
                    response = result.response
                )
            }

            is RoutingResult.PrerequisiteMissing -> {
                NavigationAction.PromptPrerequisites(
                    message = result.message,
                    missingPrerequisites = result.missing.map { it.name }
                )
            }

            is RoutingResult.ConfirmationRequired -> {
                NavigationAction.RequestConfirmation(
                    message = result.message,
                    intent = result.intent
                )
            }
        }
    }

    /**
     * Execute navigation to a target.
     * Builds route and triggers navigation via navigator.
     */
    private fun executeNavigation(result: RoutingResult.Navigate): NavigationAction {
        return try {
            val destination = deepLinkBuilder.buildDestination(result.target)
            navigator.navigate(destination)

            NavigationAction.Navigated(
                destination = destination,
                route = destination.toRoute()
            )
        } catch (e: IllegalArgumentException) {
            NavigationAction.NavigationFailed(
                error = "Invalid navigation target: ${e.message}",
                target = result.target
            )
        }
    }
}

/**
 * Result of navigation execution.
 * Returned to UI layer for appropriate feedback/state updates.
 */
sealed class NavigationAction {
    /**
     * Successfully navigated to a destination.
     */
    data class Navigated(
        val destination: NavCaddyDestination,
        val route: String
    ) : NavigationAction()

    /**
     * Display response inline without navigation.
     */
    data class ShowInlineResponse(
        val response: String
    ) : NavigationAction()

    /**
     * Prompt user to provide missing prerequisites.
     */
    data class PromptPrerequisites(
        val message: String,
        val missingPrerequisites: List<String>
    ) : NavigationAction()

    /**
     * Request user confirmation before proceeding.
     */
    data class RequestConfirmation(
        val message: String,
        val intent: caddypro.domain.navcaddy.models.ParsedIntent
    ) : NavigationAction()

    /**
     * Navigation failed due to invalid target.
     */
    data class NavigationFailed(
        val error: String,
        val target: caddypro.domain.navcaddy.models.RoutingTarget
    ) : NavigationAction()
}
